### ISSUE TITLE
Support built-in compressor for `ActiveSupport::Cache`

### ISSUE THREAD

**Issue by heka1024** [rails/rails#51700]
Rails uses `zlib` by default, but newer algorithms like `zstd` are known to show better compression and performance ([Ref](https://github.com/facebook/zstd)) We can configure compressor, but it doesn't seem to be used much due to lack of guides and the lack of other built-in compressors. 

How about separate the interface from the implementation (Zlib) now, and add zstd and another compression algorithm (maybe snappy or Lz4? just zstd for now)? I think it's better to change default compression algorithm to zstd from zlib.

And it would be nice to add a compressor with a compatible mode. My application already has huge records compressed with zlib and I don't want to break this compatibility, so I have to read the header to select the compression algorithm, like this 

```ruby
module Compressor
  ZstdCompressor = ...
  ZlibCompressor = ...

  class CombinedCompressor
    def initialize
      @zstd_compressor = ZSTDCompressor.new
      @gzip_compressor = ZlibCompressor.new
    end

    delegate :deflate, to: :@zstd_compressor # Zstd is better than Gzip

    def inflate(value)
      if @zstd_compressor.signature?(value) # Check using header value
        @zstd_compressor.inflate(value)
      elsif @gzip_compressor.signature?(value)
        @gzip_compressor.inflate(value)
      else
        Rails.logger.warn("Unknown compression signature")
        value
      end
    end
  end
end
```

To summarize, how about introduce another built-in compressor using zstd? If this is interesting, I'll create a pull request for this.

**Comment by heka1024** [comment]
@pawurb @byroot Could you give me a feedback for this issue? I found some discussions on https://github.com/rails/rails/issues/48368

**Comment by byroot** [comment]
> Rails uses zlib by default, but newer algorithms like zstd are known to show better compression and performance

Yes, Zlib isn't great, and some other options are much better. However Zlib is in the stdlib, whereas other options require a dependency on a third party gem, that have a C extension or FFI binding.

So defaulting to something else would mean increase Rails dependencies, and the risk that some of them fail to build etc. So IMO we shouldn't default to other algorithms.
